### PR TITLE
Error reporting

### LINKED ISSUE #41: Document how to use `fix` to make recursive grammars
Motivated by #40.

### LINKED ISSUE #47: Error reporting
We don’t have any right now! It’s just pass/fail. That sucks.

### PR DESCRIPTION
**Pull request by robrix**
I’m currently [experimenting in the tests](https://github.com/robrix/Madness/pull/77/files#diff-2f238e198c4816656d280b1e12a92788R3) before applying this across the interface, since it’s difficult to experiment quickly when small changes require refactoring every combinator.

An overview of the current experiment:

Error reporting replaces the `Optional` return value of parsers with `Either<Error, T>`. This means that every parser reports an error when it fails to match, describing (in brief) what it expected in order to be able to parse, but did not find.

`Error` is a simple tree, with leaves consisting of a `String` and an `IndexType`, and branches being arrays of `Error`s. These map fairly directly onto terminal/nonterminal parsers, respectively.

When implemented for all combinators, terminal parsers will each be able to directly state their unmet expectations:

```swift
let input = "abc"
(%"xyz")(input, input.startIndex) // => .left(.leaf("expected 'xyz'", input.startIndex))
```

Intuitively, nonterminals will combine their sub-parsers’ errors in some parser-dependent fashion:

```swift
let input = "abc"
(%"def" | %"geh")(input, input.startIndex) // => .left(.branch([.leaf("expected 'def'", input.startIndex), .leaf("expected 'geh'", input.startIndex)]))
```

While thorough, I fear that this will lead to poor signal to noise ratios, so there are still things to do:

- [x] Explore mixing error-reporting and non-error-reporting parsers. _Findings:_ mo’ overloads, mo’ problems.
- [x] Explore assigning human-readable identifiers to key parsers in a grammar, and using these to build error messages.
- ~~Explore maintaining and reporting a path of (identifier, index) pairs.~~ Later.
- ~~Explore reductions returning `Either` directly, enabling reporting of semantic errors at parse time.~~ Later.
- [x] Merge `master` in.

Fixes #41.
Fixes #47.

### CODE REVIEW

**Comment by robrix** [comment]
Well I guess I should also test this out a bit!

**Comment by robrix** [comment]
This is ready for review.

File: Madness/Parser.swift [hunk 1]
```diff
@@ -57,12 +57,30 @@ public prefix func % <C: CollectionType where C.Generator.Element: Equatable> (l
...
+	var memoized: T!
+	return {
+		if memoized == nil {
+			memoized = f()
+		}
+		return memoized
+	}
+}
+
+public func delay<C: CollectionType, T>(parser: () -> Parser<C, T>.Function) -> Parser<C, T>.Function {
```
Comment (robrix) [inline Madness/Parser.swift]: Apparently I’m not using this anywhere yet. Oh well.

File: Madness/Parser.swift [hunk 2]
```diff
@@ -3,31 +3,31 @@
...
+	return parser(input, input.startIndex).flatMap { $1 == input.endIndex ? .right($0) : .left(.leaf("finished parsing before end of input", $1)) }
 }
 
 
 // MARK: - Terminals
 
 /// Returns a parser which never parses its input.
-public func none<C: CollectionType, Tree>() -> Parser<C, Tree>.Function {
-	return const(nil)
+public func none<C: CollectionType, Tree>(string: String = "no way forward") -> Parser<C, Tree>.Function {
```
Comment (robrix) [inline Madness/Parser.swift]: This is a deceptively small change, so take note: it allows grammars to embed information about semantic errors!

That is, you can report better errors by detecting mistakes in the parser itself and failing via `none(…)`:

```swift
let abstraction = %"λ" ++ variable ++ (%"." | none("looks like you forgot a . !"))
```

Comment (jspahrsummers) [inline Madness/Parser.swift]: Nice!

**Comment by robrix** [comment]
Unless you have any objections I’m gonna get `master` merged in and then merge this tomorrow.

**Comment by robrix** [comment]
(The takeaway being that there’ll be issues to file on what the errors look like but I figure they can be tackled orthogonally.)

**Comment by jspahrsummers** [comment]
Not sure how soon I'll be able to return to ogdl-swift, because I'd like to get RAC 3 out the door before WWDC week. :confused:

**Comment by robrix** [comment]
That’s fine :+1:

**Pull request merged**